File: backend/src/rules/english.ts
import { GrammarRule } from '../types';

// Common Indian English errors - 50+ rules for fast detection
export const grammarRules: GrammarRule[] = [
  // Subject-Verb Agreement
  { 
    pattern: /\bI\s+has\b/gi, 
    correction: "I have", 
    type: "subject-verb", 
    explanation_hi: "'I' के साथ 'have' use करो, 'has' नहीं" 
  },
  { 
    pattern: /\b(he|she|it)\s+have\b/gi, 
    correction: "$1 has", 
    type: "subject-verb", 
    explanation_hi: "'He/She/It' के साथ 'has' use करो" 
  },
  { 
    pattern: /\bI\s+is\b/gi, 
    correction: "I am", 
    type: "subject-verb", 
    explanation_hi: "'I' के साथ 'am' use करो" 
  },
  { 
    pattern: /\b(we|they|you)\s+is\b/gi, 
    correction: "$1 are", 
    type: "subject-verb", 
    explanation_hi: "'We/They/You' के साथ 'are' use करो" 
  },
  
  // Tense Errors
  { 
    pattern: /\bdid\s+\w+ed\b/gi, 
    correction: "did + base verb", 
    type: "tense", 
    explanation_hi: "'Did' के बाद base form use करो, -ed नहीं" 
  },
  { 
    pattern: /\byesterday\s+I\s+(go|eat|come|see)\b/gi, 
    correction: "yesterday I went/ate/came/saw", 
    type: "tense", 
    explanation_hi: "Past time के लिए past tense use करो" 
  },
  { 
    pattern: /\bI\s+am\s+go\b/gi, 
    correction: "I am going / I go", 
    type: "tense", 
    explanation_hi: "'Am' के साथ '-ing' form या simple present use करो" 
  },
  { 
    pattern: /\bsince\s+\d+\s+years\b/gi, 
    correction: "for X years", 
    type: "preposition", 
    explanation_hi: "Duration के लिए 'for' use करो, 'since' specific time के लिए है" 
  },
  { 
    pattern: /\bI\s+am\s+working\s+since\b/gi, 
    correction: "I have been working since", 
    type: "tense", 
    explanation_hi: "'Since' के साथ present perfect continuous use करो" 
  },
  
  // Double negatives & Common mistakes
  { 
    pattern: /\bmore\s+better\b/gi, 
    correction: "better", 
    type: "comparative", 
    explanation_hi: "'Better' already comparative है, 'more' मत लगाओ" 
  },
  { 
    pattern: /\bhe\s+don't\b/gi, 
    correction: "he doesn't", 
    type: "negation", 
    explanation_hi: "'He/She/It' के साथ 'doesn't' use करो" 
  },
  { 
    pattern: /\bI\s+didn't\s+went\b/gi, 
    correction: "I didn't go", 
    type: "tense", 
    explanation_hi: "'Didn't' के बाद base form use करो" 
  },
  { 
    pattern: /\bdon't\s+has\b/gi, 
    correction: "don't have", 
    type: "auxiliary", 
    explanation_hi: "'Don't' के बाद base form 'have' use करो" 
  },
  
  // Fillers (instant detection)
  { 
    pattern: /\b(umm+|uhh+|aah+|uh+|hmm+)\b/gi, 
    correction: "[pause instead]", 
    type: "filler", 
    explanation_hi: "Filler sounds avoid करो, बस pause लो" 
  },
  { 
    pattern: /\b(you know)\b/gi, 
    correction: "", 
    type: "filler", 
    explanation_hi: "'You know' filler है, हटाओ" 
  },
  { 
    pattern: /\blike\s+like\b/gi, 
    correction: "like", 
    type: "filler", 
    explanation_hi: "Repeated 'like' avoid करो" 
  },
  { 
    pattern: /\bbasically\s+basically\b/gi, 
    correction: "basically", 
    type: "filler", 
    explanation_hi: "Word repeat मत करो" 
  },
  
  // Indian English specific (Indianisms)
  { 
    pattern: /\bI\s+am\s+having\s+a\s+doubt\b/gi, 
    correction: "I have a question", 
    type: "indianism", 
    explanation_hi: "English में 'doubt' का मतलब अलग है, 'question' use करो" 
  },
  { 
    pattern: /\bkindly\s+do\s+the\s+needful\b/gi, 
    correction: "please help with this", 
    type: "indianism", 
    explanation_hi: "Modern English में ऐसा नहीं बोलते" 
  },
  { 
    pattern: /\bprepone\b/gi, 
    correction: "reschedule earlier / move up", 
    type: "indianism", 
    explanation_hi: "'Prepone' standard English नहीं है" 
  },
  { 
    pattern: /\brevert\s+back\b/gi, 
    correction: "reply / get back", 
    type: "indianism", 
    explanation_hi: "'Revert' का मतलब reply नहीं है English में" 
  },
  { 
    pattern: /\bout\s+of\s+station\b/gi, 
    correction: "out of town / traveling", 
    type: "indianism", 
    explanation_hi: "'Out of station' Indian English है" 
  },
  { 
    pattern: /\bdo\s+one\s+thing\b/gi, 
    correction: "[remove or rephrase]", 
    type: "indianism", 
    explanation_hi: "Direct instruction दो, 'do one thing' filler है" 
  },
  
  // Articles
  { 
    pattern: /\bI\s+am\s+engineer\b/gi, 
    correction: "I am an engineer", 
    type: "article", 
    explanation_hi: "Profession से पहले 'a/an' लगाओ" 
  },
  { 
    pattern: /\bhe\s+is\s+doctor\b/gi, 
    correction: "he is a doctor", 
    type: "article", 
    explanation_hi: "Singular countable noun से पहले article चाहिए" 
  },
  { 
    pattern: /\bshe\s+is\s+teacher\b/gi, 
    correction: "she is a teacher", 
    type: "article", 
    explanation_hi: "Profession से पहले 'a/an' लगाओ" 
  },
  
  // Prepositions
  { 
    pattern: /\bdiscuss\s+about\b/gi, 
    correction: "discuss", 
    type: "preposition", 
    explanation_hi: "'Discuss' के बाद 'about' नहीं लगता" 
  },
  { 
    pattern: /\benter\s+into\s+the\s+room\b/gi, 
    correction: "enter the room", 
    type: "preposition", 
    explanation_hi: "'Enter' के बाद 'into' नहीं लगता" 
  },
  { 
    pattern: /\breturn\s+back\b/gi, 
    correction: "return", 
    type: "preposition", 
    explanation_hi: "'Return' में 'back' का meaning already है" 
  },
  { 
    pattern: /\bmarried\s+with\b/gi, 
    correction: "married to", 
    type: "preposition", 
    explanation_hi: "'Married' के साथ 'to' use करो, 'with' नहीं" 
  },
  
  // More common errors
  { 
    pattern: /\bI\s+am\s+agree\b/gi, 
    correction: "I agree", 
    type: "verb", 
    explanation_hi: "'Agree' simple verb है, 'am' की जरूरत नहीं" 
  },
  { 
    pattern: /\bI\s+am\s+disagree\b/gi, 
    correction: "I disagree", 
    type: "verb", 
    explanation_hi: "'Disagree' simple verb है, 'am' की जरूरत नहीं" 
  },
  { 
    pattern: /\bI\s+am\s+understand\b/gi, 
    correction: "I understand", 
    type: "verb", 
    explanation_hi: "'Understand' simple verb है, 'am' की जरूरत नहीं" 
  },
  { 
    pattern: /\bmuch\s+people\b/gi, 
    correction: "many people", 
    type: "quantifier", 
    explanation_hi: "Countable के साथ 'many' use करो, 'much' uncountable के लिए है" 
  },
  { 
    pattern: /\bmuch\s+students\b/gi, 
    correction: "many students", 
    type: "quantifier", 
    explanation_hi: "Countable के साथ 'many' use करो" 
  },
  { 
    pattern: /\bless\s+people\b/gi, 
    correction: "fewer people", 
    type: "quantifier", 
    explanation_hi: "Countable के साथ 'fewer' use करो, 'less' uncountable के लिए है" 
  },
  { 
    pattern: /\bI\s+have\s+went\b/gi, 
    correction: "I have gone / I went", 
    type: "tense", 
    explanation_hi: "'Have' के साथ past participle 'gone' use करो" 
  },
  { 
    pattern: /\bI\s+have\s+ate\b/gi, 
    correction: "I have eaten / I ate", 
    type: "tense", 
    explanation_hi: "'Have' के साथ past participle 'eaten' use करो" 
  },
  { 
    pattern: /\bI\s+have\s+saw\b/gi, 
    correction: "I have seen / I saw", 
    type: "tense", 
    explanation_hi: "'Have' के साथ past participle 'seen' use करो" 
  },
  { 
    pattern: /\bone\s+of\s+my\s+friend\b/gi, 
    correction: "one of my friends", 
    type: "plural", 
    explanation_hi: "'One of' के बाद plural form use करो" 
  },
  { 
    pattern: /\btell\s+me\s+about\s+yourself\b/gi, 
    correction: "tell me about yourself", 
    type: "common-phrase", 
    explanation_hi: "यह सही है - interview में commonly पूछा जाता है" 
  },
  { 
    pattern: /\bwhat\s+is\s+your\s+good\s+name\b/gi, 
    correction: "what is your name", 
    type: "indianism", 
    explanation_hi: "'Good name' Indian English है, सिर्फ 'name' बोलो" 
  },
  { 
    pattern: /\bI\s+am\s+doing\s+my\s+homework\s+since\s+morning\b/gi, 
    correction: "I have been doing my homework since morning", 
    type: "tense", 
    explanation_hi: "'Since' के साथ present perfect continuous use करो" 
  },
  { 
    pattern: /\bI\s+am\s+here\s+since\s+yesterday\b/gi, 
    correction: "I have been here since yesterday", 
    type: "tense", 
    explanation_hi: "'Since' के साथ present perfect use करो" 
  },
  { 
    pattern: /\bless\s+words\b/gi, 
    correction: "fewer words", 
    type: "quantifier", 
    explanation_hi: "Countable के साथ 'fewer' use करो" 
  },
  { 
    pattern: /\bI\s+done\s+my\s+work\b/gi, 
    correction: "I did my work / I have done my work", 
    type: "tense", 
    explanation_hi: "Auxiliary verb 'have' या simple past 'did' use करो" 
  },
  { 
    pattern: /\bI\s+seen\s+that\s+movie\b/gi, 
    correction: "I saw that movie / I have seen that movie", 
    type: "tense", 
    explanation_hi: "Auxiliary verb 'have' या simple past 'saw' use करो" 
  },
  { 
    pattern: /\bmore\s+easier\b/gi, 
    correction: "easier", 
    type: "comparative", 
    explanation_hi: "'Easier' already comparative है" 
  },
  { 
    pattern: /\bmore\s+faster\b/gi, 
    correction: "faster", 
    type: "comparative", 
    explanation_hi: "'Faster' already comparative है" 
  },
  { 
    pattern: /\bvery\s+much\s+good\b/gi, 
    correction: "very good", 
    type: "adverb", 
    explanation_hi: "'Much' की जरूरत नहीं, सिर्फ 'very good' बोलो" 
  },
  { 
    pattern: /\btoo\s+much\s+hot\b/gi, 
    correction: "too hot / very hot", 
    type: "adverb", 
    explanation_hi: "'Too hot' या 'very hot' बोलो" 
  },
  { 
    pattern: /\bI\s+am\s+belonging\s+to\b/gi, 
    correction: "I belong to", 
    type: "verb", 
    explanation_hi: "Stative verbs में '-ing' नहीं लगता" 
  },
  { 
    pattern: /\bI\s+am\s+knowing\b/gi, 
    correction: "I know", 
    type: "verb", 
    explanation_hi: "Stative verbs में '-ing' नहीं लगता" 
  },
];

export function detectGrammarError(text: string): { hasError: boolean; rule?: GrammarRule; match?: RegExpMatchArray } {
  for (const rule of grammarRules) {
    const match = text.match(rule.pattern);
    if (match) {
      return { hasError: true, rule, match };
    }
  }
  return { hasError: false };
}

File: backend/internal/rules/english.go
package rules

import (
	"regexp"
	"strings"
)

// GrammarRule represents a single grammar rule
type GrammarRule struct {
	ID          string
	Pattern     *regexp.Regexp
	ErrorType   string
	Description string
	Correction  func(string) string
}

// GrammarRules contains all 50+ grammar rules
var GrammarRules = []GrammarRule{
	// Subject-Verb Agreement
	{
		ID:          "I_HAS",
		Pattern:     regexp.MustCompile(`(?i)\bI has\b`),
		ErrorType:   "Subject-Verb Agreement",
		Description: "Use 'have' with 'I', not 'has'",
		Correction:  func(s string) string { return strings.ReplaceAll(strings.ToLower(s), "i has", "I have") },
	},
	{
		ID:          "HE_HAVE",
		Pattern:     regexp.MustCompile(`(?i)\b(he|she|it) have\b`),
		ErrorType:   "Subject-Verb Agreement",
		Description: "Use 'has' with 'he/she/it', not 'have'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\b(he|she|it) have\b`).ReplaceAllString(s, "$1 has") },
	},
	{
		ID:          "THEY_IS",
		Pattern:     regexp.MustCompile(`(?i)\bthey is\b`),
		ErrorType:   "Subject-Verb Agreement",
		Description: "Use 'are' with 'they', not 'is'",
		Correction:  func(s string) string { return strings.ReplaceAll(strings.ToLower(s), "they is", "they are") },
	},
	{
		ID:          "WE_WAS",
		Pattern:     regexp.MustCompile(`(?i)\bwe was\b`),
		ErrorType:   "Subject-Verb Agreement",
		Description: "Use 'were' with 'we', not 'was'",
		Correction:  func(s string) string { return strings.ReplaceAll(strings.ToLower(s), "we was", "we were") },
	},

	// Tense Errors
	{
		ID:          "YESTERDAY_GO",
		Pattern:     regexp.MustCompile(`(?i)\byesterday\b.*\bgo\b`),
		ErrorType:   "Tense Error",
		Description: "Use past tense 'went' with 'yesterday'",
		Correction:  func(s string) string { return strings.ReplaceAll(s, " go ", " went ") },
	},
	{
		ID:          "LAST_WEEK_DO",
		Pattern:     regexp.MustCompile(`(?i)\blast (week|month|year)\b.*\bdo\b`),
		ErrorType:   "Tense Error",
		Description: "Use past tense 'did' with past time markers",
		Correction:  func(s string) string { return strings.ReplaceAll(s, " do ", " did ") },
	},
	{
		ID:          "TOMORROW_WENT",
		Pattern:     regexp.MustCompile(`(?i)\btomorrow\b.*\bwent\b`),
		ErrorType:   "Tense Error",
		Description: "Use future tense with 'tomorrow'",
		Correction:  func(s string) string { return strings.ReplaceAll(s, " went ", " will go ") },
	},

	// Indianisms
	{
		ID:          "DO_THE_NEEDFUL",
		Pattern:     regexp.MustCompile(`(?i)\bdo the needful\b`),
		ErrorType:   "Indianism",
		Description: "Replace with 'please take necessary action' or 'please do what is needed'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bdo the needful\b`).ReplaceAllString(s, "please take necessary action") },
	},
	{
		ID:          "PREPONE",
		Pattern:     regexp.MustCompile(`(?i)\bprepone\b`),
		ErrorType:   "Indianism",
		Description: "Use 'reschedule earlier' or 'move forward' instead",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bprepone\b`).ReplaceAllString(s, "reschedule earlier") },
	},
	{
		ID:          "REVERT_BACK",
		Pattern:     regexp.MustCompile(`(?i)\brevert back\b`),
		ErrorType:   "Redundancy",
		Description: "'Revert' already means 'back', use just 'revert' or 'reply'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\brevert back\b`).ReplaceAllString(s, "reply") },
	},
	{
		ID:          "UPDATION",
		Pattern:     regexp.MustCompile(`(?i)\bupdation\b`),
		ErrorType:   "Indianism",
		Description: "Use 'update' instead of 'updation'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bupdation\b`).ReplaceAllString(s, "update") },
	},

	// Articles
	{
		ID:          "MISSING_ARTICLE_A",
		Pattern:     regexp.MustCompile(`(?i)\b(have|need|want|see) (book|car|house|pen)\b`),
		ErrorType:   "Missing Article",
		Description: "Add article 'a' before singular countable nouns",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\b(have|need|want|see) (book|car|house|pen)\b`).ReplaceAllString(s, "$1 a $2") },
	},
	{
		ID:          "THE_INDIA",
		Pattern:     regexp.MustCompile(`(?i)\bthe India\b`),
		ErrorType:   "Unnecessary Article",
		Description: "Don't use 'the' with country names (except USA, UK, etc.)",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bthe India\b`).ReplaceAllString(s, "India") },
	},

	// Prepositions
	{
		ID:          "DIFFERENT_THAN",
		Pattern:     regexp.MustCompile(`(?i)\bdifferent than\b`),
		ErrorType:   "Wrong Preposition",
		Description: "Use 'different from', not 'different than'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bdifferent than\b`).ReplaceAllString(s, "different from") },
	},
	{
		ID:          "MARRIED_WITH",
		Pattern:     regexp.MustCompile(`(?i)\bmarried with\b`),
		ErrorType:   "Wrong Preposition",
		Description: "Use 'married to', not 'married with'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bmarried with\b`).ReplaceAllString(s, "married to") },
	},
	{
		ID:          "DISCUSS_ABOUT",
		Pattern:     regexp.MustCompile(`(?i)\bdiscuss about\b`),
		ErrorType:   "Unnecessary Preposition",
		Description: "Use 'discuss', not 'discuss about'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bdiscuss about\b`).ReplaceAllString(s, "discuss") },
	},

	// Double Negatives
	{
		ID:          "DONT_HAVE_NOTHING",
		Pattern:     regexp.MustCompile(`(?i)\bdon't have nothing\b`),
		ErrorType:   "Double Negative",
		Description: "Use 'don't have anything' instead",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bdon't have nothing\b`).ReplaceAllString(s, "don't have anything") },
	},
	{
		ID:          "CANT_NEVER",
		Pattern:     regexp.MustCompile(`(?i)\bcan't never\b`),
		ErrorType:   "Double Negative",
		Description: "Use 'can never' instead",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bcan't never\b`).ReplaceAllString(s, "can never") },
	},

	// Fillers (common in speech)
	{
		ID:          "UMM_FILLER",
		Pattern:     regexp.MustCompile(`(?i)\b(umm|ummm|uhh|uhhh)\b`),
		ErrorType:   "Filler Word",
		Description: "Avoid using filler words like 'umm', 'uhh'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\b(umm|ummm|uhh|uhhh)\b`).ReplaceAllString(s, "") },
	},
	{
		ID:          "LIKE_FILLER",
		Pattern:     regexp.MustCompile(`(?i)\b(like)\b.*\b(like)\b.*\b(like)\b`),
		ErrorType:   "Excessive Filler",
		Description: "Reduce excessive use of 'like'",
		Correction:  func(s string) string { return s },
	},
	{
		ID:          "YOU_KNOW_FILLER",
		Pattern:     regexp.MustCompile(`(?i)\byou know\b`),
		ErrorType:   "Filler Phrase",
		Description: "Avoid filler phrase 'you know'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\byou know\b`).ReplaceAllString(s, "") },
	},

	// Plural/Singular
	{
		ID:          "THIS_THINGS",
		Pattern:     regexp.MustCompile(`(?i)\bthis (things|people|books|cars)\b`),
		ErrorType:   "Singular/Plural Mismatch",
		Description: "Use 'these' with plural nouns, not 'this'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bthis (things|people|books|cars)\b`).ReplaceAllString(s, "these $1") },
	},
	{
		ID:          "THESE_THING",
		Pattern:     regexp.MustCompile(`(?i)\bthese (thing|person|book|car)\b`),
		ErrorType:   "Singular/Plural Mismatch",
		Description: "Use 'this' with singular nouns, not 'these'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bthese (thing|person|book|car)\b`).ReplaceAllString(s, "this $1") },
	},

	// Word Order
	{
		ID:          "ALWAYS_NOT",
		Pattern:     regexp.MustCompile(`(?i)\balways not\b`),
		ErrorType:   "Word Order",
		Description: "Use 'not always' instead of 'always not'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\balways not\b`).ReplaceAllString(s, "not always") },
	},

	// Comparatives
	{
		ID:          "MORE_BETTER",
		Pattern:     regexp.MustCompile(`(?i)\bmore better\b`),
		ErrorType:   "Double Comparative",
		Description: "Use 'better', not 'more better'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bmore better\b`).ReplaceAllString(s, "better") },
	},
	{
		ID:          "MORE_WORSE",
		Pattern:     regexp.MustCompile(`(?i)\bmore worse\b`),
		ErrorType:   "Double Comparative",
		Description: "Use 'worse', not 'more worse'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bmore worse\b`).ReplaceAllString(s, "worse") },
	},

	// Common Mistakes
	{
		ID:          "COULD_OF",
		Pattern:     regexp.MustCompile(`(?i)\bcould of\b`),
		ErrorType:   "Common Mistake",
		Description: "Use 'could have' or 'could've', not 'could of'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bcould of\b`).ReplaceAllString(s, "could have") },
	},
	{
		ID:          "WOULD_OF",
		Pattern:     regexp.MustCompile(`(?i)\bwould of\b`),
		ErrorType:   "Common Mistake",
		Description: "Use 'would have' or 'would've', not 'would of'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bwould of\b`).ReplaceAllString(s, "would have") },
	},
	{
		ID:          "SHOULD_OF",
		Pattern:     regexp.MustCompile(`(?i)\bshould of\b`),
		ErrorType:   "Common Mistake",
		Description: "Use 'should have' or 'should've', not 'should of'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bshould of\b`).ReplaceAllString(s, "should have") },
	},

	// Less/Fewer
	{
		ID:          "LESS_PEOPLE",
		Pattern:     regexp.MustCompile(`(?i)\bless (people|students|items|things)\b`),
		ErrorType:   "Less vs Fewer",
		Description: "Use 'fewer' with countable nouns, not 'less'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bless (people|students|items|things)\b`).ReplaceAllString(s, "fewer $1") },
	},

	// Your/You're
	{
		ID:          "YOUR_ARE",
		Pattern:     regexp.MustCompile(`(?i)\byour (going|coming|being)\b`),
		ErrorType:   "Your vs You're",
		Description: "Use 'you're' (you are), not 'your'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\byour (going|coming|being)\b`).ReplaceAllString(s, "you're $1") },
	},

	// Their/There/They're
	{
		ID:          "THEIR_ARE",
		Pattern:     regexp.MustCompile(`(?i)\btheir are\b`),
		ErrorType:   "Their vs There",
		Description: "Use 'there are', not 'their are'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\btheir are\b`).ReplaceAllString(s, "there are") },
	},

	// Its/It's
	{
		ID:          "ITS_BEING",
		Pattern:     regexp.MustCompile(`(?i)\bits (going|coming|being)\b`),
		ErrorType:   "Its vs It's",
		Description: "Use 'it's' (it is), not 'its'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bits (going|coming|being)\b`).ReplaceAllString(s, "it's $1") },
	},

	// Then/Than
	{
		ID:          "BETTER_THEN",
		Pattern:     regexp.MustCompile(`(?i)\bbetter then\b`),
		ErrorType:   "Then vs Than",
		Description: "Use 'than' for comparisons, not 'then'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bbetter then\b`).ReplaceAllString(s, "better than") },
	},

	// Affect/Effect
	{
		ID:          "EFFECT_VERB",
		Pattern:     regexp.MustCompile(`(?i)\bwill effect\b`),
		ErrorType:   "Affect vs Effect",
		Description: "Use 'affect' as a verb, 'effect' as a noun",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bwill effect\b`).ReplaceAllString(s, "will affect") },
	},

	// More Indianisms
	{
		ID:          "OUT_OF_STATION",
		Pattern:     regexp.MustCompile(`(?i)\bout of station\b`),
		ErrorType:   "Indianism",
		Description: "Use 'out of town' instead of 'out of station'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bout of station\b`).ReplaceAllString(s, "out of town") },
	},
	{
		ID:          "PASS_OUT",
		Pattern:     regexp.MustCompile(`(?i)\bI (pass out|passed out) from college\b`),
		ErrorType:   "Indianism",
		Description: "Use 'graduate' instead of 'pass out' for education",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\b(pass out|passed out) from\b`).ReplaceAllString(s, "graduated from") },
	},
	{
		ID:          "GOOD_NAME",
		Pattern:     regexp.MustCompile(`(?i)\bgood name\b`),
		ErrorType:   "Indianism",
		Description: "Just ask 'What is your name?', not 'What is your good name?'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bgood name\b`).ReplaceAllString(s, "name") },
	},

	// More tense errors
	{
		ID:          "SINCE_PRESENT",
		Pattern:     regexp.MustCompile(`(?i)\bsince\b.*\b(go|come|work)\b`),
		ErrorType:   "Tense Error",
		Description: "Use present perfect tense with 'since'",
		Correction:  func(s string) string { return s },
	},
	{
		ID:          "FOR_PAST",
		Pattern:     regexp.MustCompile(`(?i)\bfor (two|three|four|five) (years|months|days)\b.*\bworked\b`),
		ErrorType:   "Tense Error",
		Description: "Use present perfect with duration (for/since)",
		Correction:  func(s string) string { return strings.ReplaceAll(s, "worked", "have been working") },
	},

	// Question formation
	{
		ID:          "WHERE_YOU_ARE",
		Pattern:     regexp.MustCompile(`(?i)\bwhere you are\b`),
		ErrorType:   "Question Formation",
		Description: "Use 'where are you' in questions",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bwhere you are\b`).ReplaceAllString(s, "where are you") },
	},
	{
		ID:          "WHAT_YOU_WANT",
		Pattern:     regexp.MustCompile(`(?i)\bwhat you want\b`),
		ErrorType:   "Question Formation",
		Description: "Use 'what do you want' in questions",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bwhat you want\b`).ReplaceAllString(s, "what do you want") },
	},

	// More subject-verb agreement
	{
		ID:          "EVERYONE_ARE",
		Pattern:     regexp.MustCompile(`(?i)\beveryone are\b`),
		ErrorType:   "Subject-Verb Agreement",
		Description: "'Everyone' is singular, use 'is' not 'are'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\beveryone are\b`).ReplaceAllString(s, "everyone is") },
	},
	{
		ID:          "SOMEBODY_ARE",
		Pattern:     regexp.MustCompile(`(?i)\b(somebody|someone|anybody|anyone) are\b`),
		ErrorType:   "Subject-Verb Agreement",
		Description: "Indefinite pronouns are singular, use 'is'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\b(somebody|someone|anybody|anyone) are\b`).ReplaceAllString(s, "$1 is") },
	},

	// Redundancies
	{
		ID:          "REPEAT_AGAIN",
		Pattern:     regexp.MustCompile(`(?i)\brepeat again\b`),
		ErrorType:   "Redundancy",
		Description: "'Repeat' already means 'again', just use 'repeat'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\brepeat again\b`).ReplaceAllString(s, "repeat") },
	},
	{
		ID:          "RETURN_BACK",
		Pattern:     regexp.MustCompile(`(?i)\breturn back\b`),
		ErrorType:   "Redundancy",
		Description: "'Return' already means 'back', just use 'return'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\breturn back\b`).ReplaceAllString(s, "return") },
	},

	// More common errors
	{
		ID:          "ALOT",
		Pattern:     regexp.MustCompile(`(?i)\balot\b`),
		ErrorType:   "Spelling Error",
		Description: "Use 'a lot' (two words), not 'alot'",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\balot\b`).ReplaceAllString(s, "a lot") },
	},
	{
		ID:          "CANT_ABLE_TO",
		Pattern:     regexp.MustCompile(`(?i)\bcan't able to\b`),
		ErrorType:   "Double Modal",
		Description: "Use either 'can't' or 'not able to', not both",
		Correction:  func(s string) string { return regexp.MustCompile(`(?i)\bcan't able to\b`).ReplaceAllString(s, "am not able to") },
	},
}

// DetectError checks text against all grammar rules
func DetectError(text string) (*GrammarRule, string) {
	text = strings.TrimSpace(text)
	if text == "" {
		return nil, ""
	}

	for i := range GrammarRules {
		rule := &GrammarRules[i]
		if rule.Pattern.MatchString(text) {
			corrected := rule.Correction(text)
			return rule, corrected
		}
	}

	return nil, ""
}
